### PR TITLE
Agentic Training

### PR DESCRIPTION
**Pull request by tastelikefeet**
# PR type
- [x] Bug Fix
- [x] New Feature
- [x] Document Updates
- [ ] More Models or Datasets Support

# PR information

1. **CP gradient normalization fix** — Divide `num_tokens` by `cp_size` in `reduce_loss`, fixing the bug where grad_norm scales inversely with cp_size when CP > 1

2. **Megatron embedding task support** — Add `_resolve_task_context` context manager and `MegatronEmbeddingPatch` (output layer swap + CP gather + last-token pooling) to enable contrastive embedding training

3. **InfoNCE contrastive loss** — Add `InfoNCELoss` and `EmbeddingMetric` for a complete embedding training loss + evaluation pipeline

4. **Megatron forward improvements** — Pass `cu_seqlens` to `postprocess_tensor_cp` for correct CP reconstruction with packing; conditional logps computation (gated on `require_logps`); grad_norm type fix

5. **Tool-call parsing framework** — Add `template/tools/` module supporting `cline`/`react`/`vcp`/`qwen` tool-call format parsers; Qwen3 chat template patch to relax tool-tail restrictions

6. **Agentic data preprocessing pipeline** — Add 15 new filters: dead_loop, dedup, hard, refuse, PII, intent classifier, message_sanity, token_soup, score, etc., for agent trajectory data cleaning

7. **Router sampler** — Add `router_sampler.py` to support routing-based sampling request dispatch

8. **CLI module** — Add `twinkle/cli/` providing dataset export and management command-line operations

9. **Dataset extensions** — Add `save_as()` export functionality in `base.py`; processor CP handling with cu_seqlens support

10. **Cookbook scripts** — Add condenser training, embedding training, and cold-start training scripts; archive legacy GRPO scripts; update fsdp2 cookbook

11. **Test coverage** — Add 12 new test files covering CLI, dataset export, 8 preprocessor filters, and tool-call parsers

12. **Infrastructure** — Dockerfile update, pyproject.toml version bump, new utility functions in `utils/parallel.py` and `torch_utils.py`, `no_split_modules` patch

## Experiment results

Paste your experiment result here(if needed).

### CODE REVIEW

File: src/twinkle/loss/infonce.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    """
+    split_indices = torch.nonzero(labels, as_tuple=False).squeeze().tolist()
+    if isinstance(split_indices, int):
+        split_indices = [split_indices]
+    split_indices.append(len(labels))
+    split_indices = np.array(split_indices) + np.array(list(range(len(split_indices))))
+    split_tensors = []
+    for i in range(len(split_indices) - 1):
+        start, end = split_indices[i], split_indices[i + 1]
+        split_part = sentences[start:end]
```
[the remark this answers is not in the document]
Comment (tastelikefeet) [reply src/twinkle/loss/infonce.py:70]: solved

**Review by hjh0119 — APPROVED** [review]

**Review by Jintao-Huang — APPROVED** [review]

File: src/twinkle/loss/infonce.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+        local_label_shape = labels.new_tensor(labels.shape, dtype=torch.long)
+        label_shapes = [torch.empty_like(local_label_shape) for _ in range(world_size)]
+        dist.all_gather(label_shapes, local_label_shape, group=self.process_group)
+        all_labels = [labels.new_empty(shape.tolist()) for shape in label_shapes]
+        dist.all_gather(all_labels, labels.contiguous(), group=self.process_group)
+
+        # keep the local shard differentiable; detach others
+        all_sentences[rank] = sentences
+        for idx in range(world_size):
```
[the remark this answers is not in the document]
Comment (tastelikefeet) [reply src/twinkle/loss/infonce.py:155]: 不会出现长度不同的情况
